docs/04-refactor-state-management.md

```diff
@@ -91,126 +90,125 @@ public void ResetOrder()
+@functions {
+    [Parameter] Action Clicked { get; set; }
+}
 
-To hook up the other side, we should subscribe to the change notifications on `Index`
+@* TestComponent3.razor *@
+@inject TestState State
+<CoolButton Clicked="@State.Clicked" />
+<p>Clicked @State.Count times!</p>
 ```
-@implements IDisposable
 
-...
+In this third example the event handler delegate is `TestState.Clicked` and the so `Delegate.Target` is `TestState` - **not a component**. When the button is clicked, no component gets the event notification, and so nothing will rerender.
 
-@functions {
-    ...
+This is the problem that `EventCallback` was created to solve. By changing the parameter on `CoolButton` from `Action` to `EventCallback` you fix the event dispatching behavior. This works because `EventCallback` is known to the compiler, when you create an `EventCallback` from a delegate that doesn't have its `Target` set to a component, then the compiler will pass the curent component to recieve the event.
 
-    protected async override Task OnInitAsync()
-    {
-        OrderState.StateChanged += OnOrderStateChanged;
-        specials = await HttpClient.GetJsonAsync<List<PizzaSpecial>>("specials");
-    }
+----
 
-    void IDisposable.Dispose()
-    {
-        OrderState.StateChanged -= OnOrderStateChanged;
-    }
+Let's jump back to our application. If you like you can reproduce the problem that's been described here by changing the parameters of `ConfigurePizzaDialog` from `EventCallback` to `Action`. If you try this you can see that cancelling or confirming the dialog does nothing. This is because our use case is exactly like the third example above:
 
-    void OnOrderStateChanged(object sender, EventArgs e) => StateHasChanged();
+```html
+@* from Index.razor *@
+@if (OrderState.ShowingConfigureDialog)
+{
+    <ConfigurePizzaDialog
+        Pizza="@OrderState.ConfiguringPizza"
+        OnConfirm="@OrderState.ConfirmConfigurePizzaDialog" 
+        OnCancel="@OrderState.CancelConfigurePizzaDialog" />
 }
 ```
 
-This is just a normal .NET event handler, and the `Index` component will unsubscribe from changes once it gets disposed.
-
-note: You might be wondering whether it's necessary to make `ConfigurePizzaDialog` subscribe to `OrderState.StateChanged` too. As it happens, there's no need in this case. That's because `ConfigurePizzaDialog` only ever needs to re-render in response to its own DOM events, and the framework automatically triggers a re-render for you in that case. This is different from `Index`, because `Index` also needs to re-render after `ConfigurePizzaDialog` has modified `OrderState`.
+For the `OnConfirm` and `OnCancel` parameters, the `Delegate.Target` will be `OrderState` since we're passing reference to methods defined by `OrderState`. If you're using `EventCallback` then the special logic of the compiler kicks in and it will specify additional information to dispatch the event to `Index`.
```

Comment:
I'm concerned that we're spending an excessive amount of time on `EventCallback` here. I know it's an important concept, but there are so many other important concepts too. Maybe we could only mention it once and not come back to it again here.